src/device.cc

```diff
@@ -64,6 +64,30 @@ bool DeviceConnection::removeSubDevice(const QString& path)
   return false;
 }
 
+// -------------------------------------------------------------------------------------------------
+void DeviceConnection::queryBatteryStatus()
+{
+  if (subDeviceCount() > 0) {
+    for (const auto& sd: subDevices()) {
+      if (sd.second->type() == ConnectionType::Hidraw && sd.second->mode() == ConnectionMode::ReadWrite) {
```

Comment:
But don't worry about refactoring for possible (custom) hidraw communication with other devices (than the Logitech Spotlight) - I will think about this after the Logitech features have been implemented.

Reply:
For this commit, I have added the flag for battery. However, there is a better way to enquire this support from the device. From the comments in `queryBatteryStatus` in (device.cc L383)
```
  // if we make battery feature request packet by sending {0x11, 0x01, 0x00, 0x0d, 0x10, 0x00 ... padded by 0x00}
  // batteryFeatureID is the fifth byte obtained by as device response.
  // batteryFeatureID may differ for different logitech devices and may change after firmware update.
  // last checked, batteryFeatureID was 0x06 for logitech spotlight.
```
This logic can be applied to all supported features in Logitech. In total the device support 27 features. I am able to get following feature table from device:
```
Feature Table for Spotlight (27 Features Total)
Feature Index in device (uint8_t) : Universal Feature Code in HID++2.0 (uint16_t): Description
0: 0000 : Root Feature
1: 0001 : Feature Set
2: 0003 : Device Firmware Version
3: 0005 : Device Name
4: 1D4B : Wireless Device Status
5: 0020 : Reset
6: 1000 : Battery
7: 1B04 : Reprogram Controls V4
8: 2005 : SWAP_BUTTON_CANCEL,
9: 1A00 : Presenter Control
0x0a: 2205 : Pointer Speed
0x0b: 1870 : Unknown (Software_Hidden)
0x0c: 1A01 : Sensor 3D
0x0d: 00C2 : DFUCONTROL_SIGNED
0x0e: 1803 : Unknown   (Software_Hidden)
0x0f: 1805 : OOBSTATE  (Software_Hidden)
0x10: 1806 : Config Device Properties  (Software_Hidden)
0x11: 1811 : Unknown   (Software_Hidden)
0x12: 1830 : Unknown   (Software_Hidden)
0x13: 1890 : Unknown   (Software_Hidden)
0x14: 1891 : Unknown   (Software_Hidden)
0x15: 18A1 : Unknown   (Software_Hidden)
0x16: 1DF3 : Unknown   (Software_Hidden)
0x17: 1E00 : Unknown   (Software_Hidden)
0x18: 1EB0 : Unknown   (Software_Hidden)
0x19: 1861 : Unknown   (Software_Hidden)
0x1a: 1850 : Unknown   (Software_Hidden)
```

By using this logic, we can make the projecteur work for all possible firmware update in the future. I am working on it and will submit a PR to be included in v1.0.

Reply:
Hi @jahnf ,

I have added the required code. Now the software read the features supported by device and the enable the right flags (see `initSubDevice` function in `device.cc`).

I am not free for next 3-4 weeks. There are two issues with current code. I have listed them, in case you wish to review or work on it:

1. This reading of feature set is not done at beginning if the device was not active. As soon as any key on device is pressed, the reading of feature set and setting of device flag is done. However, the device tab in preferences dialog box is not getting updated with new device flags as the dialog was initialized before. Can you look into it? If the device tab is updated whenever the Preferences dialog is shown (`ShowEvent`) then it might be fixed.

2. Additionally, can you please check the possibility of saving the firmware details and feature set in setting cache file? (Please see `hidpp.cc` L 57-59).

Reply:
Thank you very much, will definitely work on integrating until end of July. 👍